github-actions/commit-message-based-labels/lib/main.ts

```diff
@@ -11,66 +12,107 @@ const supportedLabels = [
 
-  try {
+class CommitMessageBasedLabelManager {
+  /** Run the commit message based labelling process. */
+  static run = async () => {
     const token = await getAuthTokenFor(ANGULAR_ROBOT);
-    installationClient = new Octokit({auth: token});
+    const git = new Octokit({auth: token});
+    const inst = new this(git);
+    await inst.run();
+  };
 
-    await runCommitMessageBasedLabelsAction(installationClient);
-  } finally {
-    if (installationClient !== null) {
-      await revokeActiveInstallationToken(installationClient);
-    }
-  }
-}
-
-async function runCommitMessageBasedLabelsAction(client: Octokit): Promise<void> {
-  const {number, owner, repo} = context.issue;
   /** Labels currently applied to the PR. */
-  const labels = await (
-    await client.issues.listLabelsOnIssue({issue_number: number, owner, repo})
-  ).data;
+  labels = new Set<string>();
   /** Parsed commit message for every commit on the PR. */
-  const commits = await (
-    await client.paginate(client.pulls.listCommits, {owner, pull_number: number, repo})
-  ).map(({commit: {message}}) => parseCommitMessage(message));
+  commits = new Set<Commit>();
 
-  console.log(`PR #${number}`);
+  private constructor(private git: Octokit) {}
 
-  // Add or Remove label as appropriate for each of the supported label and commit messaage
-  // combinations.
-  for (const [label, commitProperty] of supportedLabels) {
-    const hasCommit = commits.some((commit) => commit[commitProperty].length > 0);
-    const hasLabel = labels.some(({name}) => name === label);
-    console.log(`${commitProperty} | hasLabel: ${hasLabel} | hasCommit: ${hasCommit}`);
+  /** Run the action, and revoke the installation token on completion. */
+  async run() {
+    try {
+      // Initialize the labels and commits before performing the action.
+      await this.initialize();
+      core.info(`PR #${context.issue.number}`);
 
-    if (hasCommit && !hasLabel) {
-      await client.issues.addLabels({
-        repo,
-        owner,
-        issue_number: number,
-        labels: [label],
-      });
-      console.log(`Added ${label} label to PR #${number}`);
+      // Add or Remove label as appropriate for each of the supported label and commit messaage
+      // combinations.
+      for (const [label, commitProperty] of supportedLabels) {
+        const hasCommit = [...this.commits].some((commit) => commit[commitProperty].length > 0);
+        const hasLabel = this.labels.has(label);
+        core.info(`${commitProperty} | hasLabel: ${hasLabel} | hasCommit: ${hasCommit}`);
+
+        if (hasCommit && !hasLabel) {
+          await this.addLabel(label);
+        }
+        if (!hasCommit && hasLabel) {
+          await this.removeLabel(label);
+        }
+      }
+
+      // Add 'comp: docs' label for changes which contain a docs commit.
+      if (!this.labels.has(compDocsLabel)) {
+        for (const commit of this.commits) {
+          if (commit.type === COMMIT_TYPES['docs'].name) {
```

Comment:
Any reason for the docs being accessed through an index signature? It would be more type-safe using plain property access expressions

Reply:
Because the type of `COMMIT_TYPES` is `{[key: string}: CommitType}` so it can't be regular property access from typescripts perspective.

Reply:
It would be possible to make it a constant though. Especially since we don't support custom commit types in `ng-dev`. I'm good with this as is, can be a future clean-up, but it's worth noting that currently this is "unsafe"

Reply:
In the worst case we change the key from `docs` and we wouldn't even get a type error, but a runtime error because we read `.name` from `undefined`.